src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Char.cs

```diff
@@ -646,10 +646,10 @@ public static unsafe int IndexOfNullCharacter(char* searchSpace)
                             Debug.Assert(lengthToExamine >= Vector512<ushort>.Count);
 
                             Vector512<ushort> search = *(Vector512<ushort>*)(searchSpace + (nuint)offset);
-                            ulong matches = Vector512.Equals(Vector512<ushort>.Zero, search).AsByte().ExtractMostSignificantBits();
+
                             // Note that MoveMask has converted the equal vector elements into a set of bit flags,
                             // So the bit position in 'matches' corresponds to the element offset.
-                            if (matches == 0)
+                            if (!Vector512.EqualsAny(search, Vector512<ushort>.Zero))
```

Comment:
The pattern here is, in general, more closely similar to how we handle the `V128` path to support `Arm64`.

Since `Arm64` doesn't have `pmovmsk`, we typically do a cheaper check against `zero` first and then extract the mask only when we have some match.

In the same way, AVX-512 doesn't make it trivial to extract a mask directly and so its cheaper to do the check against zero and then get an actual mask only when we have a match.

There's potentially some CSE/VN improvements we could do around the pattern here, but in general this ends up working well overall and produces much better codegen.

Reply:
hhm, shouldn't this be `Equals()` instead? At least that's what we do for Arm64.

Also, the comments should be updated because there is no `matches` anymore. Also, if you don't mind, can you update the comment in Arm64 case too that says `"// Zero flags set so no matches"`.

Reply:
`Equals(x, y)` and `EqualsAny(x, y)` aren't quite the same. One returns a vector and the other a bool, the latter is itself effectively `Equals(x, y) != Vector<T>.Zero`

In the case of AVX-512 (and it will also be true for SVE) we don't get a vector from the underlying hardware, we get a mask and so this ends up being better for the common case of "no match", at the cost of 1 additional instruction for exit condition. Particularly for long inputs, this ends up being significantly better. That is, the actual codegen for `Equals(x, y)` is `vpcmpeq; vpmovm2*` and for `EqualsAny` is `vpcmpeq; kortest`

In .NET 9 we'll likely want to update things so that `Equals` and `EqualsAny` can have the shared `vpcmpeq` CSE'd, but its a bit too complex and not as impactful for .NET 9 since it primarily only impacts 512-bit codegen.

Reply:
Fixed the comment.